### ISSUE TITLE
Devmode should be recognizing compile-only and runtime-only dependencies

### ISSUE THREAD

**Issue by aloubyansky** [quarkusio/quarkus#9422]
Gradle supports `compileOnly` and `runtimeOnly` classpath dependencies. However, at this point our devmode support does not allow isolating `compileOnly` dependencies from the runtime classpath.
And isolate the `runtimeOnly` classpath from the compile classpath.

fyi @stuartwdouglas

**Comment by stuartwdouglas** [comment]
Maven has the same thing (provided = compileOnly, runtime=runtimeOnly). 

I actually just ran into an issue caused by this with my reagument PR, we are resolving provided deps which includes graal svm, which includes a bundled old version of ASM.

If I get time today I might make an attempt at this based on your PR for the 'test dep leakage' issue.

**Comment by aloubyansky** [comment]
I thought about Maven and `provided` opening this issue too. It's not the same. `provided` is (mis)used for that. But what provided means is "use the dependency for compilation only, it will be provided at runtime by the environment". In that case we should be including `provided` for running the app. I am not sure the promise is the same for `compileOnly` in Gradle.

**Comment by stuartwdouglas** [comment]
I don't think we should be supplying 'provided', otherwise from the point of view of Quarkus it is the same as 'compile'.

We also use provided for artifacts that we expect substrate VM to provide, which definitely should not go into the app.

**Comment by aloubyansky** [comment]
You are right, I was just thinking the same. So, yes, we should not included `provided`.

**Comment by FWest98** [comment]
We are running into this issue, where we'd like to have some dependencies as `compileOnly` in Gradle, but Quarkus doesn't like this in dev mode...

**Comment by aloubyansky** [comment]
As `provided` in Maven? Keep in mind that `provided` dependencies are added to the test runtime classpath by Maven itself (and Quarkus too).

**Comment by FWest98** [comment]
I use Gradle as follows:
```kotlin
compileOnly("foo.bar")
testImplementation("foo.bar")
```

For non-Quarkus projects, this works perfectly fine. However, when trying to run the tests in the Quarkus project, I get:
```
Execution failed for task ':project:quarkusGenerateCode'.
> Could not resolve all dependencies for configuration ':project:quarkusProdCompileOnlyConfiguration'.
   > Could not find com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:.
     Required by:
         project :project
```

**Comment by aloubyansky** [comment]
Could you please create a new issue and attach a reproducer? Thanks

**Comment by FWest98** [comment]
Turns out I was running into https://github.com/quarkusio/quarkus/issues/39073, so a different issue than this.

**Comment by geoand** [comment]
@aloubyansky is this still relevent?

**Comment by aloubyansky** [comment]
I think with https://github.com/quarkusio/quarkus/pull/51261 we could consider excluding Gradle's `compileOnly` from the runtime dependencies of an app in dev mode.
`runtimeOnly` would require more work to separate.
OTOH, it doesn't seem like a critical issue and we haven't had many complaints. So, let's close it for now.

**Comment by geoand** [comment]
👍🏽
